### PR TITLE
Fix file/URI parsing with archive syntax

### PR DESCRIPTION
**Pull request by antonio2368**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Correctly parse file name/URI containing `::` if it's not an archive. 

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

File: src/IO/S3/URI.cpp [hunk 1]
```diff
@@ -54,10 +55,7 @@ URI::URI(const std::string & uri_)
     static constexpr auto OSS = "OSS";
     static constexpr auto EOS = "EOS";
 
-    if (containsArchive(uri_))
-        std::tie(uri_str, archive_pattern) = getPathToArchiveAndArchivePattern(uri_);
-    else
-        uri_str = uri_;
+    std::tie(uri_str, archive_pattern) = getURIAndArchivePattern(uri_);
```
Comment (antonio2368) [inline src/IO/S3/URI.cpp:58]: cc @divanik

**Review by alexey-milovidov — APPROVED** [review]

File: tests/queries/0_stateless/03215_parsing_archive_name_s3.sql [hunk 2]
```diff
@@ -0,0 +1,6 @@
+-- Tags: no-fasttest
+-- Tag no-fasttest: Depends on AWS
+
+SELECT _file, _path FROM s3(s3_conn, filename='::03215_archive.csv') ORDER BY (_file, _path);
```
Comment (alexey-milovidov) [inline tests/queries/0_stateless/03215_parsing_archive_name_s3.sql:4]: Also add at least one example with `file`

Comment (antonio2368) [reply tests/queries/0_stateless/03215_parsing_archive_name_s3.sql:4]: check `tests/queries/0_stateless/03214_parsing_archive_name_file.sh`

I'm also adding a setting `allow_archive_path_syntax`

**Review by divanik — APPROVED** [review]

**Comment by divanik** [comment]
Could you also add to the s3 table function docs in reading from archive section that user needs to assure that allow_archive_path_syntax is enabled?

**Comment by helifu** [comment]
It's invalid to have "::" in the file name on WINDOWS system:

```
$ git rebase upstream/master
error: invalid path 'tests/queries/0_stateless/data_minio/::03215_archive.csv'
error: invalid path 'tests/queries/0_stateless/data_minio/test.zip::03215_archive.csv'
error: invalid path 'tests/queries/0_stateless/data_minio/test::03215_archive.csv'
error: could not detach HEAD
...

$ git config core.protectNTFS false
$ git rebase upstream/master
error: unable to create file tests/queries/0_stateless/data_minio/::03215_archive.csv: Invalid argument
error: unable to create file tests/queries/0_stateless/data_minio/test.zip::03215_archive.csv: Invalid argument
error: unable to create file tests/queries/0_stateless/data_minio/test::03215_archive.csv: Invalid argument
error: could not detach HEAD
```
Is there any way to work around it? :(

**Comment by alexey-milovidov** [comment]
@antonio2368 let's rename the files in the Git repository and rename them back in tests.

**Comment by antonio2368** [comment]
@helifu thanks for reporting and sorry for the issue!
the fix is here https://github.com/ClickHouse/ClickHouse/pull/67882

File: src/Core/SettingsChangesHistory.cpp [hunk 3]
```diff
@@ -76,6 +76,7 @@ static std::initializer_list<std::pair<ClickHouseVersion, SettingsChangesHistory
     {"24.8",
         {
             {"merge_tree_min_bytes_per_task_for_remote_reading", 4194304, 2097152, "Value is unified with `filesystem_prefetch_min_bytes_for_single_read_task`"},
+            {"allow_archive_path_syntax", true, true, "Added new setting to allow disabling archive path syntax."},
```
Comment (rschu1ze) [inline src/Core/SettingsChangesHistory.cpp:79]: @antonio2368 L. 79 / 155 seem weird. Should that be

```cpp
/// 24.8
{"allow_archive_path_syntax", false, true, "Added new setting to allow disabling archive path syntax."},

/// 24.5
{"allow_archive_path_syntax", false, false, "Added new setting to allow disabling archive path syntax."},
```

?

**Pull request merged**